hazelcast/src/test/java/com/hazelcast/client/alto/ClientAltoTest.java

```diff
@@ -0,0 +1,501 @@
+        Config config = getMemberConfig();
+        Hazelcast.newHazelcastInstance(config);
+
+        HazelcastInstance client = HazelcastClient.newHazelcastClient(getClientConfig());
+        ClientConnectionManager connectionManager = getConnectionManager(client);
+        Collection<ClientConnection> connections = connectionManager.getActiveConnections();
+
+        assertClientConnectsAllAltoPortsEventually(connections, config.getAltoConfig().getEventloopCount());
+
+        CountDownLatch disconnected = new CountDownLatch(1);
+        CountDownLatch reconnected = new CountDownLatch(1);
+        client.getLifecycleService().addLifecycleListener(state -> {
+            if (state.getState() == LifecycleEvent.LifecycleState.CLIENT_DISCONNECTED) {
+                disconnected.countDown();
+            } else if (state.getState() == LifecycleEvent.LifecycleState.CLIENT_CONNECTED) {
+                reconnected.countDown();
+            }
+        });
+
+        assertEquals(1, connections.size());
+        TcpClientConnection connection = (TcpClientConnection) connections.iterator().next();
+        Channel[] channels = connection.getAltoChannels();
+
+        connection.close("Expected", null);
+
+        assertOpenEventually(disconnected);
+
+        assertFalse(connection.isAlive());
+        for (Channel channel : channels) {
+            // All the channels must be closed as well
+            assertTrue(channel.isClosed());
+        }
+
+        assertOpenEventually(reconnected);
+
+        assertClientConnectsAllAltoPortsEventually(connections, config.getAltoConfig().getEventloopCount());
+    }
+
+    @Test
+    public void testPartitionBoundPendingInvocations_whenConnectionCloses() {
+        String mapName = randomMapName();
+        Config config = getMemberWithMapStoreConfig(mapName);
+        Hazelcast.newHazelcastInstance(config);
+
+        HazelcastInstance client = HazelcastClient.newHazelcastClient(getClientConfig());
+        ClientConnectionManager connectionManager = getConnectionManager(client);
+        Collection<ClientConnection> connections = connectionManager.getActiveConnections();
+
+        assertClientConnectsAllAltoPortsEventually(connections, config.getAltoConfig().getEventloopCount());
+
+        IMap<Integer, Integer> map = client.getMap(mapName);
+        CompletableFuture<Integer> future = map.putAsync(1, 1).toCompletableFuture();
+        connections.iterator().next().close("Expected", null);
+
+        // Should get TargetDisconnectedException and retried based on our rules,
+        // which bubbles the exception to the user for non-retryable messages
+        assertThatThrownBy(future::join)
+                .isInstanceOf(CompletionException.class)
+                .hasCauseInstanceOf(TargetDisconnectedException.class);
+    }
+
+    @Test
+    public void testAltoChannelTargetedPendingInvocations_whenConnectionCloses() {
+        // We don't send invocations to Alto channels this way, but this is just
```

Comment:
how do we send them ? did not see any releated changes in the latest commit. Also, invokeOnPartitionOwner code does not seem to sending to alto channels on this branch. Can you remind me where the related code was?

Reply:
It is handled in the TcpClientConnection#write